apollo-api/src/main/java/com/spotify/apollo/HeadersValue.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * -\-\-
+ * Spotify Apollo API Interfaces
+ * --
+ * Copyright (C) 2013 - 2016 Spotify AB
+ * --
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * -/-/-
+ */
+package com.spotify.apollo;
+
+import com.google.auto.value.AutoValue;
+import com.google.common.base.CharMatcher;
+
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Function;
+
+import javax.annotation.Nonnull;
+
+/**
+ * Ensures that the lower-case contract of {@link Headers} is met by the value objects.
+ */
+@AutoValue
+abstract class HeadersValue implements Headers {
+
+  private static final CharMatcher ASCII_UPPER_CASE = CharMatcher.inRange('A', 'Z');
+
+  @Nonnull
+  @Override
+  public Optional<String> get(@Nonnull String name) {
+    return Optional.ofNullable(asMap().get(name.toLowerCase()));
+  }
+
+  static Headers create(Map<String, String> headers) {
+    Map<String, String> allLowerCaseKeys = headers.keySet().stream()
+        .filter(HeadersValue::hasNonLowerCase)
+        .findAny()
+        .map(lowerCaseKeys(headers))
+        .orElse(headers);
```

Comment:
this stream operation seems confusing to me:

1. it would seem to filter out any keys that are all lower-case 
2. and the map operation ignores the input (the `s` in `s ->`)

is the goal to avoid modifying the map in the happy case when the keys are already all lower case?